### PR TITLE
Decorrelate subquery with left UNNEST and global aggregation

### PR DESCRIPTION
**Pull request by kasiafi**

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateLeftUnnestWithGlobalAggregation.java [hunk 1]
```diff
@@ -0,0 +1,277 @@
+/*
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateLeftUnnestWithGlobalAggregation.java:1]: How much common code is here and in https://github.com/trinodb/trino/pull/6326. Should those be merged maybe?

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateLeftUnnestWithGlobalAggregation.java [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+ *           agg <- agg1(x)
+ *           - Projection
+ *                x <- foo(y)
+ *                - Aggregation
+ *                     group by (g, a, c1, c2, unique)
+ *                     y <- agg2(u)
+ *                     - Unnest LEFT
+ *                          g <- unnest(c1)
+ *                          u <- unnest(c2)
+ *                          replicate: (a, c1, c2, unique)
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateLeftUnnestWithGlobalAggregation.java:93]: why do we need c1, c2 here?

Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateLeftUnnestWithGlobalAggregation.java:93]: Oh - we need those. As they are also part of `CorrelatedJoinNode` which was not obvious to me. 
Probably can not be part of grouping set. But maybe it does not matter.

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateLeftUnnestWithGlobalAggregation.java [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+                    input,
+                    Assignments.builder()
+                            .putIdentities(input.getOutputSymbols())
+                            .putAll(sourceProjection.getAssignments())
+                            .build());
+        }
+
+        // rewrite correlated join to UnnestNode
+        UnnestNode rewrittenUnnest = new UnnestNode(
+                context.getIdAllocator().getNextId(),
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateLeftUnnestWithGlobalAggregation.java:161]: nit: Maybe you can reuse `unnestNode.getId` here and then there is no need to pass `PlanNodeId correlatedUnnestId` to `rewriteNodeSequence`

File: core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestDecorrelateLeftUnnestWithGlobalAggregation.java [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+                                                Optional.empty(),
+                                                LEFT,
+                                                Optional.empty(),
+                                                project(
+                                                        ImmutableMap.of("char_array", expression("regexp_extract_all(corr, '.')")),
+                                                        assignUniqueId("unique", values("corr")))))));
+    }
+
+    @Test
+    public void testMultipleNodesOverUnnestInSubquery()
```
Comment (losipiuk) [inline core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestDecorrelateLeftUnnestWithGlobalAggregation.java:324]: Thanks :) I just hope no one will ever need to edit this one :p

**Review by losipiuk — COMMENTED** [review]
LGTM module possibly extracting some common stuff used here and in #6326. And maybe some other rules too?

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateLeftUnnestWithGlobalAggregation.java (hunk 2 above)
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateLeftUnnestWithGlobalAggregation.java:93]: NVM: https://github.com/trinodb/trino/pull/6326/files#r580313959

Sorry for thinking out loud :)

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateLeftUnnestWithGlobalAggregation.java (hunk 1 above)
Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateLeftUnnestWithGlobalAggregation.java:1]: They only share the tiniest helper methods. The rest of the code looks similar but differs in details. I'd rather not merge them because the code might be hard to follow. Nor put them in a single file.

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateLeftUnnestWithGlobalAggregation.java (hunk 3 above)
Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateLeftUnnestWithGlobalAggregation.java:161]: I can reuse the `PlanNodeId`, but I think that it should be the `CorrelatedJoinNode`'s id rather than `UnnestNode`'s id.
Anyway, I think that it would still be necessary to pass `PlanNodeId correlatedUnnestId`. Could you explain how I could avoid that?

File: core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestDecorrelateLeftUnnestWithGlobalAggregation.java (hunk 4 above)
Comment (kasiafi) [reply core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestDecorrelateLeftUnnestWithGlobalAggregation.java:324]: Oh. I never thought that tests might get in the way of future progress. If they do, I'll be there to help.

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateLeftUnnestWithGlobalAggregation.java (hunk 3 above)
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateLeftUnnestWithGlobalAggregation.java]: If we reused `UnnestNode` id it would not be necessary. As you could do:
```java
        // rewrite correlated join to UnnestNode
        UnnestNode rewrittenUnnest = new UnnestNode(
                unnestNode.getId(),
                input,
                input.getOutputSymbols(),
                unnestNode.getMappings(),
                unnestNode.getOrdinalitySymbol(),
                LEFT,
                Optional.empty());

        // restore all projections, grouped aggregations and global aggregations from the subquery
        PlanNode result = rewriteNodeSequence(
                context.getLookup().resolve(correlatedJoinNode.getSubquery()),
                input.getOutputSymbols(),
                rewrittenUnnest,
                context.getLookup());
```

and then
```java
    private static PlanNode rewriteNodeSequence(PlanNode root, List<Symbol> leftOutputs, PlanNode sequenceSource, Lookup lookup)
    {
        // bottom of sequence reached -- attach the rewritten source node
        if (root.getId().equals(sequenceSource.getId())) {
            return sequenceSource;
        }
```

But it is not anything super important.

**Review by losipiuk — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateLeftUnnestWithGlobalAggregation.java (hunk 3 above)
Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateLeftUnnestWithGlobalAggregation.java]: Right. But I think it might be confusing: having two nodes with the same id at the same time.
And also, `rewrittenUnnest` is the logical successor of `correlatedJoinNode` rather than of `unnestNode`.

Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateLeftUnnestWithGlobalAggregation.java]: I'm leaning towards not reusing PlanNodeId here at all.

Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateLeftUnnestWithGlobalAggregation.java]: Sure. Keep it.
